sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -151,7 +150,8 @@ private Bounded(@Nullable String name, BoundedSource<T> source) {
           .apply(ParDo.of(new OutputSingleSource<>(source)))
           .setCoder(SerializableCoder.of(new TypeDescriptor<BoundedSource<T>>() {}))
           .apply(ParDo.of(new BoundedSourceAsSDFWrapperFn<>()))
-          .setCoder(source.getOutputCoder());
+          .setCoder(source.getOutputCoder())
+          .setTypeDescriptor(source.getOutputCoder().getEncodedTypeDescriptor());
```

Comment:
Do we maintain the `TypeDescriptor` information before? I was under impression that for most of cases we only set `Coder` for a output `PCollection`.

Reply:
You are right and I don't know why we don't pay more attention to this. Probably because coders seem to include the TypeDescriptor, any ideas @kennknowles ? is this redundant somehow?

In any case having this information seems important for the downstream transforms.

Reply:
It seems like the `typeDescriptor` can be inferred from `Coder.getEncodedTypeDescriptor()`. If we really want to populate this information in a consistent way, probably we can consider changing `PCollection.getTypeDescriptor()` to infer the `typeDescriptor` from `Coder` if the `typeDescriptor` is set.

Reply:
Yes, the expected use of this method is to set the type descriptor but _not_ the coder. This way, the coder registry still can choose the coder.

Setting both is redundant, in theory. Setting just the coder should suffice. Maybe some plumbing needed? It was not really expected to look at either one in this way.

Another angle to consider is that type descriptor is Java-specific, while coder is the portable "type" of the data. I don't know if that matters here.

Reply:
I'm thinking about changes like: https://github.com/apache/beam/pull/14870

Reply:
I like @boyuanzz fix because even in the presence of different Coders the TypeDescriptor is commonly preserved inside of the Coders. WDYT @kennknowles can you spot some particular issues about it?
I can rebase this PR targetting a generic implementation like the one on #14870 but I did not do it like that because I was not really familiar with the reasoning behind not relying on the coder typeDescriptor.

Reply:
Yea makes lots of sense.